### PR TITLE
[GPU] Fix RemoteBlob lock() and ulock() behaviour in case of multiple threads

### PR DESCRIPTION
**Pull request by sshlyapn**
### Details:
This PR fixes data races in RemoteBlob implementation 

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/remote_context.cpp [hunk 1]
```diff
@@ -189,14 +189,22 @@ void RemoteBlobImpl::lock() const {
...
         IE_THROW(NotAllocated) << "[GPU] Remote blob can't be locked as it's not allocated";
     }
-    lockedHolder = std::unique_ptr<cldnn::mem_lock<uint8_t>>(new cldnn::mem_lock<uint8_t>(m_memObject, m_stream));
-    auto ptr = lockedHolder->data();
-    _handle = reinterpret_cast<void*>(ptr);
-    m_allocator.regLockedBlob(_handle, this);
+
+    std::lock_guard<std::mutex> locker(lockedMutex);
+    if (lockedCounter == 0) {
+        lockedHolder = std::unique_ptr<cldnn::mem_lock<uint8_t>>(new cldnn::mem_lock<uint8_t>(m_memObject, m_stream));
```
Comment (myshevts) [inline src/plugins/intel_gpu/src/plugin/remote_context.cpp:195]: gpu_buffer::lock() that implements the cldnn::mem_lock also has a lock

**Review by tgubanova-lohika — CHANGES_REQUESTED** [review]
Original comment was removed because it violates [A Code of Conduct for Open Source Communities](https://www.contributor-covenant.org/version/2/1/code_of_conduct/)

**Review by myshevts — APPROVED** [review]

File: src/plugins/intel_gpu/src/plugin/remote_context.cpp (hunk 1 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/plugin/remote_context.cpp:195]: Discussed offline
Short summary: it's okay to have mutex here for extra synchronization for RemoteBlob's lock()/unlock() calls

**Review by vladimir-paramuzov — APPROVED** [review]

File: src/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp [hunk 2]
```diff
@@ -98,6 +99,64 @@ TEST_P(RemoteBlob_Test, smoke_canInputUserBlob) {
...
+    GTEST_SKIP();
+#endif
+    const int batch = 2;
+    const int channels = 3;
+    const int height = 512;
+    const int width = 512;
+    const size_t img_size = batch * channels * height * width;
+    cl_int err;
+
+    const InferenceEngine::TensorDesc tensor_desc{InferenceEngine::Precision::U8,
```
Comment (iefode) [inline src/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp:113]: Do you plan to migrate these tests to OV2.0?

**Review by iefode — APPROVED** [review]

File: src/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp (hunk 2 above)
Comment (vladimir-paramuzov) [reply src/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp:113]: This test is specific to legacy API. New API doesn't allow to lock/unclock remote tensors.

**Review by apankratovantonp — APPROVED** [review]

**Pull request merged**